sdk/storage/azure-storage-common/inc/azure/storage/common/storage_per_operation_policy.hpp

```diff
@@ -0,0 +1,34 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+#pragma once
+
+#include <memory>
+#include <string>
+
+#include <azure/core/http/policy.hpp>
+
+namespace Azure { namespace Storage { namespace Details {
+
+  class StoragePerOperationPolicy : public Core::Http::HttpPolicy {
```

Comment:
> StoragePerOperationPolicy this name looks like a general purpose policy? And it is currently setting the apiVersion but it can be used to do anything?
I would prefer single-responsibility policies and renaming this policy to StorageApiVersionPolicy.

Yes, it is a general purpose policy, just it only sets API version for now, more functionalities may be added in the future.

It looks very like another policy `StoragePerRetryPolicy`, the name of which is suggested by Jeffrey. We only set `x-ms-date` in that policy, for now.

Reply:
> Also, the pattern we use to expose policy options to an SDK client is to create another structure like StoragePerOperationPolicyOptions where all defaults are defined. Then you would use that struct as part of the SDK Client options. See how it is done with the TransportPolicyOptions and RetryPolicyOptions.

Current impl of adding `std::string ApiVersion` is also suggested by Jeffrey, which I also believe is the ideal way for our customers to change the API version if they want. It's more convenience and straightforward compared with something like `clientOptions.PerOperationPolicyOptions.ApiVersion = blablabla.`.

Reply:
This name does seem strange to me. Did I like this name?
If the policy changes the api version, then the policy name should reflect what it does.

Reply:
@Jinming-Hu    Yeah, I kep thinking about this. It is fine not to have the options structure because this policy is only for storage.

The advantage of the policy options pattern is for policies from core that are used by multiple clients. In that case, if core adds a new option, all clients are updated to support that right away.

So, never mind about that. Only worry about the name. Consider a better name for your policy and avoid having general purpose policies

Reply:
@JeffreyRichter not sure if I got it wrong https://github.com/Azure/azure-sdk-for-cpp/pull/1349#discussion_r556745774

Reply:
Your implementation is right; it's the name of the policy class that is in question (the other names are right).
I think the best thing to do is this: In Azure Core, we add a ValuesPolicy whose constructor takes a ValuePolicyOptions which has a vector of a structure that has 3 fields: string key, string value, flag indicating query parameter or header. 
The ValuePolicy ctor takes the ValuePolicyOptions and the policy's Send method walks the vector and appends each query parameter to the URL or each header to the HTTP requests before calling the next policy.

The Storage clients add this policy to the header of the pipeline and initialize the ValuePolicyOptions vector with just 1 entry with "x-ms-version", the customer's value (or default), and flag indicating header.

Most Azure services use an "api-version" query parameter set to the version value. This same policy can be used for multiple Azure services and for multiple query parameters/values.

Reply:
> Your implementation is right; it's the name of the policy class that is in question (the other names are right).
I think the best thing to do is this: In Azure Core, we add a ValuesPolicy whose constructor takes a ValuePolicyOptions which has a vector of a structure that has 3 fields: string key, string value, flag indicating query parameter or header.
The ValuePolicy ctor takes the ValuePolicyOptions and the policy's Send method walks the vector and appends each query parameter to the URL or each header to the HTTP requests before calling the next policy.

> The Storage clients add this policy to the header of the pipeline and initialize the ValuePolicyOptions vector with just 1 entry with "x-ms-version", the customer's value (or default), and flag indicating header.

> Most Azure services use an "api-version" query parameter set to the version value. This same policy can be used for multiple Azure services and for multiple query parameters/values.

I think this'll work. Although it sounds to me kind of over-abstraction, maybe? But I'm fine with it.

@vhvb1989 Hi, can you work on this `ValuePolicy` in Core?

Reply:
Well, I like the valuePolicy in Core as a concept.  But it still seems to general purpose to me.

Usually, the policy-options structure is part of the client-options.
Would this means that, for setting/overriding the version, or any other header/query-parameter, the end-user need to know about the header/query-name and set it?

If the valuePolicyOptions is not part of the clientOptions, and the client needs to make the mapping from some particular option (like a string version option) to one entry for the valueOptions vector, then how is that different from letting the  client setting the header/query directly?

I would prefer to create 2 policies in core:
- QueryVersionPolicy
- HeaderVersionPolicy

Both policies could derive from a base policy VersionPolicy and just implement the exact behavior. Both would take  either a simple string with the version or a VersionPolicyOptions that contains the version.

This approach would let each client to define how it sets the version and will limit the scope of the policy to a well-defined operation.

I will talk to @JeffreyRichter and the Core team about this.

Reply:
Why is it `Version`? It's a general purpose policy that's used to set HTTP headers, not just `x-ms-version` header.

Reply:
> It's a general purpose policy

That's what I would like us to avoid.  I think that having 2 ways of doing the same thing is bad, and this is what a general-purpose policy enables. Now we can, for instance, set the request Id using the ValuePolicy or using the RequestIdPolicy.  And the same thing for setting the auth token header.

Reply:
> That's what I would like us to avoid. I think that having 2 ways of doing the same thing is bad, and this is what a general-purpose policy enables. Now we can, for instance, set the request Id using the ValuePolicy or using the RequestIdPolicy. And the same thing for setting the auth token header.

No, I don't think so. These policies are different. Value policy adds a definite header value, but `RequestIdPolicy` and auth policy add a dynamic value.